packages/hub/src/graphql/queries/search.ts

```diff
@@ -0,0 +1,126 @@
+import { builder } from "@/graphql/builder";
+import { prisma } from "@/prisma";
+import { Searchable } from "../types/Searchable";
+import { Searchable as PrismaSearchable } from "@prisma/client";
+import { resolveOffsetConnection } from "@pothos/plugin-relay";
+
+type SearchableWithEdgeData = PrismaSearchable & {
+  rank: number;
+  slugSnippet: string;
+  textSnippet: string;
+};
+
+const SearchEdge = builder.edgeObject({
+  type: Searchable,
+  name: "SearchEdge",
+  fields: (t) => ({
+    rank: t.float({
+      resolve: (obj) => {
+        const rank = (obj.node as SearchableWithEdgeData).rank;
+        return rank ?? 1e9;
+      },
+    }),
+    slugSnippet: t.string({
+      resolve: (obj) => {
+        const snippet = (obj.node as SearchableWithEdgeData).slugSnippet;
+        return String(snippet ?? "");
+      },
+    }),
+    textSnippet: t.string({
+      resolve: (obj) => {
+        const snippet = (obj.node as SearchableWithEdgeData).textSnippet;
+        return String(snippet ?? "");
+      },
+    }),
+  }),
+});
+
+builder.queryField("search", (t) =>
+  t.connection(
+    {
+      type: Searchable,
+      args: {
+        text: t.arg.string({ required: true }),
+      },
+      errors: {},
+      resolve: async (_, args, { session }) => {
+        return await resolveOffsetConnection(
+          { args },
+          async ({ limit, offset }) => {
+            const { text } = args;
+            return await prisma.$queryRaw<SearchableWithEdgeData[]>`
+        SELECT
+          "Searchable".*,
+          -- ranking function
+          ts_rank_cd(
+            setweight(to_tsvector(coalesce("Model".slug, '')), 'A') ||
+            setweight(to_tsvector(coalesce("SquiggleSnippet".code, '')), 'B') ||
```

Comment:
Note that `code` weight is lower than `slug` weights. This means that if there's a model named `foo` and another model that contains a variable named `foo`, they will be ranked correctly.

Reply:
I'd recommend leaving these comments in the code, instead of in github.

Reply:
I mentioned why I didn't do this in https://github.com/quantified-uncertainty/squiggle/pull/2674#discussion_r1421071255; I don't think it's a good idea to modify this query if you haven't read through relevant Postgres docs, and after that the comments won't be very useful, docs explain it better than I could.

(another very minor reason is that any inline comments here will be sent to the DB on each query over network, so they're different from the usual code comments)